middle_end/flambda2/reaper/rebuild.ml

```diff
@@ -0,0 +1,479 @@
+            in
+            let bound_static, group = List.split bound_and_group in
+            let static_const_or_code = function
+              | Deleted_code -> Static_const_or_code.deleted_code
+              | Code
+                  { params_and_body;
+                    code_metadata;
+                    free_names_of_params_and_body
+                  } ->
+                let is_my_closure_used =
+                  is_var_used env params_and_body.my_closure
+                in
+                let params_and_body =
+                  rebuild_function_params_and_body kinds env params_and_body
+                in
+                let code_metadata =
+                  if Bool.equal is_my_closure_used
+                       (Code_metadata.is_my_closure_used code_metadata)
+                  then code_metadata
+                  else (
+                    assert (not is_my_closure_used);
+                    Code_metadata.with_is_my_closure_used is_my_closure_used
+                      code_metadata)
+                in
+                let code =
+                  Code.create_with_metadata ~params_and_body ~code_metadata
+                    ~free_names_of_params_and_body
+                in
+                all_code := Code_id.Map.add (Code.code_id code) code !all_code;
+                Static_const_or_code.create_code code
+              | Static_const static_const ->
+                Static_const_or_code.create_static_const static_const
+            in
+            let group =
+              Static_const_group.create (List.map static_const_or_code group)
+            in
+            ( Bound_pattern.static (Bound_static.create bound_static),
+              Named.create_static_consts group )
+          | Set_of_closures { value_slots; alloc_mode; function_decls } ->
+            let bound =
+              match let_.bound_pattern with
+              | Set_of_closures s -> s
+              | Static _ | Singleton _ -> assert false
+            in
+            let set_of_closures =
+              rewrite_set_of_closures bound env value_slots alloc_mode
+                function_decls
+            in
+            let is_phantom =
+              Name_mode.is_phantom @@ Bound_pattern.name_mode let_.bound_pattern
+            in
+            all_slot_offsets
+              := Slot_offsets.add_set_of_closures !all_slot_offsets ~is_phantom
+                   set_of_closures;
+            let_.bound_pattern, Named.create_set_of_closures set_of_closures
+        in
+        let defining_expr = rewrite_named kinds env defining_expr in
+        RE.create_let bp defining_expr ~body:hole
+      in
+      rebuild_holed kinds env let_.parent subexpr
+    in
+    match let_.bound_pattern with
+    | Set_of_closures _ -> default ()
+    | Static _ -> default ()
+    | Singleton v ->
+      let v = Bound_var.var v in
+      if is_var_used env v then default () else erase ())
+  | Let_cont { cont; parent; handler } ->
+    let cont_handler =
```

Comment:
TODO: remove unused values from parameters

Reply:
This is more complex than anticipated due to fixed-arity continuations for the return of functions, putting a CR instead.